plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,569 @@
+    len :: Int
+    len = BS.length bs
+
+testBitByteString :: ByteString -> Integer -> Emitter (EvaluationResult Bool)
+testBitByteString bs i
+  | i < 0 || i >= bitLen = indexOutOfBoundsError "testBitByteString" bitLen i
+  | otherwise = pure . pure . dangerousRead bs $ i
+  where
+    bitLen :: Integer
+    bitLen = fromIntegral $ BS.length bs * 8
+
+{-# NOINLINE writeBitByteString #-}
+writeBitByteString :: ByteString -> Integer -> Bool -> Emitter (EvaluationResult ByteString)
+writeBitByteString bs i b
+  | i < 0 || i >= bitLen = indexOutOfBoundsError "writeBitByteString" bitLen i
+  -- When we write a bit at a location, we have to return a new copy of the
+  -- original with the bit modified. We do this as follows:
+  --
+  -- 1. Compute the byte that has to change. Because _byte_ indexes and _bit_
+  -- indexes go in opposite directions, we have to compute the byte by a
+  -- combination of modulus and offset from the end.
+  -- 2. Use the remainder to construct a mask which 'selects' the bit within the
+  -- byte we want to change.
+  -- 3. Memcpy everything over.
+  -- 4. Use the mask at the computed byte index to modify the result in-place:
+  -- we do a different operation depending on whether we're setting or clearing.
+  --
+  -- We use memcpy plus a single write as this is _much_ faster than going
+  -- byte-by-byte and checking if we've reached the index we want each time:
+  -- memcpy is highly-optimized using SIMD instructions on every platform, and a
+  -- branchy per-byte loop is absolutely horrid everywhere for speed due to the
+  -- branch count.
+  | otherwise = do
+      let (bigOffset, smallOffset) = i `quotRem` 8
+      let bigIx = fromIntegral $ byteLen - bigOffset - 1
+      let mask = bit 0 `shiftL` fromIntegral smallOffset
+      pure . pure . unsafeDupablePerformIO . unsafeUseAsCStringLen bs $ go bigIx mask
+  where
+    byteLen :: Integer
+    byteLen = fromIntegral . BS.length $ bs
+    bitLen :: Integer
+    bitLen = byteLen * 8
+    go :: Int -> Word8 -> (Ptr CChar, Int) -> IO ByteString
+    go bigIx mask (src, len) = do
+      dst <- mallocBytes len
+      void . memcpy dst src . fromIntegral $ len
+      byte :: Word8 <- peek . plusPtr src $ bigIx
+      let byte' = if b then mask .|. byte else complement mask .&. byte
+      poke (castPtr . plusPtr dst $ bigIx) byte'
+      unsafePackMallocCStringLen (dst, len)
+
+integerToByteString :: Integer -> ByteString
+integerToByteString i = case signum i of
+  0    -> BS.singleton zeroBits
+  (-1) -> twosCompToNegative . fromList . go . abs $ i
+  _    -> fromList . go $ i
+  where
+    -- We encode into Word8-sized 'limbs', using a stack to ensure that their
+    -- ordering is little-endian. Effectively, we encode as a base-256 number,
+    -- where the least significant digit is at the end.
+    go :: Integer -> [Word8]
+    go = \case
+      0 -> []
+      pos -> case pos `quotRem` 256 of
+        (d, r) -> go d <> [fromIntegral r]
```

Comment:
This is quadratic, no? You'd be better off with a DList or prepending with a final reverse at the end.

Reply:
Or maybe you could use a `Builder` and then maybe the monoid operations won't be expensive, not sure.